File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 147354. 테이블 해시 함수/kotlin/Solution.kt
import java.util.Arrays

class Solution {
    fun solution(data: Array<IntArray>, col: Int, row_begin: Int, row_end: Int): Int {
        val table = Array(data.size) { i -> Tuple(data[i], col)}
        Arrays.sort(table)
        val si = IntArray(data.size + 1)

        for (i in table.indices) {
            si[i+1] = table[i].si(i+1)
        }
        var answer = si[row_begin]
        for (i in row_begin until row_end) {
            answer = answer.xor(si[i+1])
        }
        return answer
    }
}

class Tuple(
    private val arr: IntArray,
    private val col: Int,
) : Comparable<Tuple> {

    private val id: Int = arr[0]

    fun si(i: Int): Int {
        var si = 0
        for (item in arr) {
            si += item%i
        }
        return si
    }

    override fun compareTo(other: Tuple): Int {
        if (arr[col-1] == other.arr[col-1]) {
            return other.id - this.id
        }
        return arr[col-1] - other.arr[col-1]
    }

}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 147354. 테이블 해시 함수/python/solution.py
def solution(data, col, row_begin, row_end):
    data.sort(key=lambda t: (t[col - 1], -t[0]))  # col 번째 = col-1 번 인덱스

    # row_begin -1 행(인덱스)부터 row_end -1 행(인덱스)까지
    # 각 칼럼을 i+1(인덱스 + 1 = x번째)값으로 나눈 값의 합을 순서대로 리스트화
    sums = [sum(map(lambda x: x % (i + 1), data[i])) for i in range(row_begin - 1, row_end)]

    answer = 0
    for value in sums:
        answer ^= value  # xor 연산
    return answer
